<!DOCTYPE html>

<html lang="en">
<head>
<meta charset="utf-8"/>
<meta content="IE=edge" http-equiv="X-UA-Compatible"/>
<meta content="width=device-width, initial-scale=1.0" name="viewport"/>
<link href="https://booiljung.github.io/programming/dart/chapter_09/0902/" rel="canonical"/>
<link href="../../../../img/favicon.ico" rel="shortcut icon"/>
<meta content="width=device-width, initial-scale=1.0, maximum-scale=1.0, user-scalable=0" name="viewport">
<title>try-catch-finally 문법 - 소프트웨어 융합</title>
<link href="../../../../css/bootstrap-3.3.7.min.css" rel="stylesheet"/>
<link href="../../../../css/font-awesome-4.7.0.css" rel="stylesheet"/>
<link href="../../../../css/base.css" rel="stylesheet"/>
<link href="../../../../css/highlight.css" rel="stylesheet"/>
<link href="../../../../css/custom.css" rel="stylesheet"/>
<!-- HTML5 shim and Respond.js IE8 support of HTML5 elements and media queries -->
<!--[if lt IE 9]>
        <script src="https://oss.maxcdn.com/libs/html5shiv/3.7.0/html5shiv.js"></script>
        <script src="https://oss.maxcdn.com/libs/respond.js/1.3.0/respond.min.js"></script>
    <![endif]-->
<script src="../../../../js/jquery-3.2.1.min.js"></script>
<script src="../../../../js/bootstrap-3.3.7.min.js"></script>
<script src="../../../../js/highlight.pack.js"></script>
<base target="_top"/>
<script>
      var base_url = '../../../..';
      var is_top_frame = false;
        
        var pageToc = [
          {title: "try \ube14\ub85d", url: "#_top", children: [
          ]},
          {title: "catch \ube14\ub85d", url: "#catch", children: [
          ]},
          {title: "finally \ube14\ub85d", url: "#finally", children: [
          ]},
          {title: "\uc804\uccb4 \uc608\uc2dc", url: "#_1", children: [
              {title: "\uc608\uc678 \ud750\ub984 \ub2e4\uc774\uc5b4\uadf8\ub7a8", url: "#_2" },
              {title: "\uc5ec\ub7ec \uac1c\uc758 catch \ube14\ub85d", url: "#catch_1" },
              {title: "rethrow\ub97c \ud1b5\ud55c \uc608\uc678 \uc7ac\ubc1c\uc0dd", url: "#rethrow" },
              {title: "\ube44\ub3d9\uae30 \ud568\uc218\uc5d0\uc11c\uc758 \uc608\uc678 \ucc98\ub9ac", url: "#_3" },
              {title: "\uc608\uc678 \ucc98\ub9ac\uc758 \uc7a5\uc810", url: "#_4" },
              {title: "\uc0ac\uc6a9\uc790 \uc815\uc758 \uc608\uc678", url: "#_5" },
              {title: "\uc608\uc678 \uac1d\uccb4\uc758 \ucd94\uac00 \uc815\ubcf4 \ud3ec\ud568", url: "#_6" },
          ]},
        ];

    </script>
<script src="../../../../js/base.js"></script>
<script src="../../../../js/google_analytics.js"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.7/MathJax.js?config=TeX-MML-AM_CHTML"></script>
<script src="https://www.googletagmanager.com/gtag/js?id=G-3F4LHCTF88"></script>
</meta></head>
<body>
<script>
if (is_top_frame) { $('body').addClass('wm-top-page'); }
</script>
<div class="container-fluid wm-page-content">
<a name="_top"></a>
<div aria-label="navigation" class="row wm-article-nav-buttons" role="navigation">
<div class="wm-article-nav pull-right">
<a class="btn btn-xs btn-default pull-right" href="../0903/">
        Next
        <i aria-hidden="true" class="fa fa-chevron-right"></i>
</a>
<a class="btn btn-xs btn-link" href="../0903/">
        사용자 정의 예외
      </a>
</div>
<div class="wm-article-nav">
<a class="btn btn-xs btn-default pull-left" href="../0901/">
<i aria-hidden="true" class="fa fa-chevron-left"></i>
        Previous</a><a class="btn btn-xs btn-link" href="../0901/">
        예외의 정의와 발생
      </a>
</div>
</div>
<p>Dart 언어에서 예외 처리는 프로그램이 실행 중에 발생할 수 있는 오류나 예외 상황을 처리하기 위해 사용된다. 예외는 예상치 못한 상황을 처리할 수 있는 메커니즘을 제공한다. 예외가 발생했을 때, 프로그램이 바로 종료되는 대신 예외를 처리할 수 있는 코드를 작성함으로써 프로그램의 안정성을 높일 수 있다.</p>
<h2 id="try">try 블록</h2>
<p><code>try</code> 블록은 예외가 발생할 수 있는 코드를 감싸는 데 사용된다. <code>try</code> 블록 내에서 발생한 예외는 <code>catch</code> 또는 <code>finally</code> 블록에서 처리할 수 있다. <code>try</code> 블록이 완료되면, <code>catch</code> 또는 <code>finally</code> 블록이 실행된다.</p>
<pre><code class="language-dart">try {
  // 예외가 발생할 가능성이 있는 코드
}
</code></pre>
<h2 id="catch">catch 블록</h2>
<p><code>catch</code> 블록은 <code>try</code> 블록에서 발생한 예외를 처리한다. 예외가 발생하면, 프로그램의 흐름이 <code>try</code> 블록을 중단하고 즉시 <code>catch</code> 블록으로 이동한다. 예외 객체를 사용하여 어떤 종류의 예외가 발생했는지 확인할 수 있다. Dart에서는 <code>catch</code> 블록에 예외 객체와 스택 추적 정보를 받을 수 있다.</p>
<pre><code class="language-dart">catch (e) {
  // 예외 처리 코드
}
</code></pre>
<p>또한, 예외 객체와 스택 추적 정보를 모두 받을 수도 있다.</p>
<pre><code class="language-dart">catch (e, s) {
  // 예외 처리 코드
  print('예외: $e');
  print('스택 추적: $s');
}
</code></pre>
<h2 id="finally">finally 블록</h2>
<p><code>finally</code> 블록은 예외가 발생하든 발생하지 않든 항상 실행되는 코드이다. 예외가 발생하더라도 반드시 실행되어야 할 코드가 있을 때 유용하다. 예를 들어, 파일을 열었다면, 파일을 닫는 작업을 반드시 해야 하는데, 이때 <code>finally</code> 블록에서 해당 작업을 처리할 수 있다.</p>
<pre><code class="language-dart">finally {
  // 예외 발생 여부와 상관없이 항상 실행되는 코드
}
</code></pre>
<p>이러한 <code>finally</code> 블록은 자원 해제 작업을 포함한 후처리 작업에 주로 사용된다.</p>
<h2 id="_1">전체 예시</h2>
<p>다음은 <code>try</code>, <code>catch</code>, <code>finally</code> 블록을 모두 사용하는 예시이다:</p>
<pre><code class="language-dart">void main() {
  try {
    int result = 12 ~/ 0;
    print(result);
  } catch (e) {
    print('예외 발생: $e');
  } finally {
    print('이 코드는 항상 실행된다.');
  }
}
</code></pre>
<p>이 코드에서 <code>12 ~/ 0</code>은 정수 나눗셈이므로, 0으로 나누는 예외가 발생하여 <code>catch</code> 블록으로 이동한다. <code>catch</code> 블록에서 예외를 출력한 후 <code>finally</code> 블록이 실행된다.</p>
<h3 id="_2">예외 흐름 다이어그램</h3>
<p><code>try-catch-finally</code>의 동작 흐름을 다이어그램으로 나타내면 다음과 같다:</p>
<div class="mermaid">graph TD;
    A[try 블록 실행] --&gt; B{예외 발생 여부};
    B -- 예외 발생 --&gt; C[catch 블록 실행];
    B -- 예외 없음 --&gt; D[finally 블록 실행];
    C --&gt; D;
    D --&gt; E[프로그램 계속 실행];
</div>
<h3 id="catch_1">여러 개의 <code>catch</code> 블록</h3>
<p>Dart에서는 <code>catch</code> 블록을 여러 개 사용할 수 있으며, 각 블록은 다른 예외 유형을 처리하도록 지정할 수 있다. 이를 통해 더 세부적으로 예외를 처리할 수 있다. Dart의 모든 예외는 <code>Exception</code> 클래스 또는 그 하위 클래스의 인스턴스로 처리된다.</p>
<pre><code class="language-dart">try {
  // 예외 발생 가능 코드
} on FormatException {
  // 특정 예외 유형 처리 (FormatException)
} catch (e) {
  // 다른 모든 예외 처리
}
</code></pre>
<p>위 코드는 <code>FormatException</code> 예외가 발생하면 해당 예외를 처리하는 <code>on</code> 블록이 실행되고, 그 외의 다른 예외는 <code>catch</code> 블록에서 처리된다.</p>
<p><code>on</code> 키워드를 사용하면 예외 유형을 명확하게 지정할 수 있다. 반면, <code>catch</code>는 모든 예외를 처리하는데 적합한다.</p>
<h3 id="rethrow"><code>rethrow</code>를 통한 예외 재발생</h3>
<p>때로는 <code>catch</code> 블록에서 예외를 처리한 후, 그 예외를 다시 던져 다른 상위 코드에서 추가로 처리해야 할 때가 있다. 이때 <code>rethrow</code> 키워드를 사용하여 예외를 재발생시킬 수 있다.</p>
<pre><code class="language-dart">try {
  // 예외 발생 가능 코드
} catch (e) {
  print('예외 처리 중: $e');
  rethrow; // 예외를 다시 던짐
}
</code></pre>
<p>이 코드는 예외를 한 번 처리한 후, 그 예외를 상위 호출 스택으로 다시 던져 상위 코드에서 추가 처리가 가능하도록 한다.</p>
<h3 id="_3">비동기 함수에서의 예외 처리</h3>
<p>비동기 함수에서도 동일한 방식으로 예외 처리를 할 수 있지만, <code>Future</code>와 함께 사용될 경우에는 <code>await</code>와 <code>try-catch</code>를 결합해야 한다. 비동기 함수에서 예외가 발생할 경우, 해당 예외는 <code>Future</code> 객체에 전달되기 때문에 <code>try-catch</code> 블록에서 이를 처리해야 한다.</p>
<pre><code class="language-dart">Future&lt;void&gt; asyncFunction() async {
  try {
    await someAsyncTask();
  } catch (e) {
    print('비동기 작업 중 예외 발생: $e');
  } finally {
    print('비동기 작업 종료');
  }
}
</code></pre>
<p>비동기 함수에서 발생한 예외도 일반적인 <code>catch</code> 블록에서 처리할 수 있으며, <code>finally</code> 블록은 여전히 작업 완료 후에 실행된다.</p>
<h3 id="_4">예외 처리의 장점</h3>
<p>예외 처리는 프로그램의 오류를 우아하게 처리할 수 있도록 하며, 프로그램이 예외로 인해 중단되지 않고 계속 실행되도록 도와준다. 예외 처리를 사용하지 않을 경우, 프로그램은 예기치 않은 오류로 인해 강제 종료될 수 있지만, <code>try-catch-finally</code> 구조를 사용하면 예외 상황을 처리하면서도 프로그램의 흐름을 제어할 수 있다.</p>
<p>다음으로 Dart의 예외 처리에서 유용하게 사용되는 사용자 정의 예외와 예외의 다양한 활용 방법에 대해 다룰 것이다.</p>
<h3 id="_5">사용자 정의 예외</h3>
<p>Dart에서는 기본적으로 제공되는 예외 외에도 개발자가 직접 예외를 정의할 수 있다. 이를 통해 특정 상황에서 발생할 수 있는 예외를 세분화하고, 더 명확하게 예외를 처리할 수 있다.</p>
<p>사용자 정의 예외는 <code>Exception</code> 클래스를 상속받아 정의할 수 있다. <code>Exception</code> 클래스는 Dart에서 예외를 처리하는 데 사용되는 기본 클래스이다. 사용자 정의 예외 클래스는 주로 예외 메시지나 추가적인 예외 정보를 담는 필드를 포함할 수 있다.</p>
<pre><code class="language-dart">class CustomException implements Exception {
  final String message;

  CustomException(this.message);

  @override
  String toString() =&gt; 'CustomException: $message';
}
</code></pre>
<p>위의 예제는 <code>CustomException</code>이라는 사용자 정의 예외를 정의하는 코드이다. <code>toString()</code> 메소드를 재정의하여 예외가 발생했을 때 출력될 메시지를 지정할 수 있다.</p>
<p>사용자 정의 예외를 사용하여 예외를 발생시키는 예는 다음과 같다.</p>
<pre><code class="language-dart">void validateInput(int input) {
  if (input &lt; 0) {
    throw CustomException('입력 값은 음수가 될 수 없다.');
  }
}
</code></pre>
<p><code>throw</code> 키워드를 사용하여 예외를 발생시킬 수 있으며, 이 예외는 <code>try-catch</code> 블록으로 처리된다.</p>
<pre><code class="language-dart">void main() {
  try {
    validateInput(-1);
  } catch (e) {
    print(e);
  }
}
</code></pre>
<p>이 코드는 <code>validateInput(-1)</code>을 호출하여 예외를 발생시키고, <code>catch</code> 블록에서 해당 예외를 처리한다. 출력 결과는 다음과 같이 나타난다.</p>
<pre><code>CustomException: 입력 값은 음수가 될 수 없다.
</code></pre>
<p>이처럼 사용자 정의 예외는 특정 도메인이나 비즈니스 로직에 맞는 예외 처리를 가능하게 하여, 코드의 가독성을 높이고, 문제를 보다 쉽게 디버깅할 수 있다.</p>
<h3 id="_6">예외 객체의 추가 정보 포함</h3>
<p>때로는 예외 객체에 추가적인 정보를 포함시켜야 할 때가 있다. Dart에서는 이를 위해 예외 객체에 필요한 데이터를 전달할 수 있다. 예를 들어, 오류 코드나 관련 데이터를 함께 제공함으로써 예외 상황을 더 구체적으로 설명할 수 있다.</p>
<pre><code class="language-dart">class DetailedException implements Exception {
  final String message;
  final int errorCode;

  DetailedException(this.message, this.errorCode);

  @override
  String toString() =&gt; 'Error $errorCode:$message';
}
</code></pre>
<p>이 예제에서는 <code>errorCode</code>라는 필드를 추가하여 예외의 구체적인 오류 코드를 함께 전달할 수 있다. 이를 통해 예외 상황을 더욱 구체적으로 파악할 수 있다.</p>
<pre><code class="language-dart">void main() {
  try {
    throw DetailedException('파일을 찾을 수 없다.', 404);
  } catch (e) {
    print(e);
  }
}
</code></pre>
<p>이 코드는 다음과 같은 결과를 출력한다:</p>
<pre><code>Error 404: 파일을 찾을 수 없다.
</code></pre>
<p>이와 같은 방식을 사용하면 예외 처리 시 추가적인 정보를 제공할 수 있어, 문제를 디버깅하고 해결하는 데 도움이 된다.</p>
<br/>
<div aria-label="navigation" class="row wm-article-nav-buttons" role="navigation">
<div class="wm-article-nav pull-right">
<a class="btn btn-xs btn-default pull-right" href="../0903/">
        Next
        <i aria-hidden="true" class="fa fa-chevron-right"></i>
</a>
<a class="btn btn-xs btn-link" href="../0903/">
        사용자 정의 예외
      </a>
</div>
<div class="wm-article-nav">
<a class="btn btn-xs btn-default pull-left" href="../0901/">
<i aria-hidden="true" class="fa fa-chevron-left"></i>
        Previous</a><a class="btn btn-xs btn-link" href="../0901/">
        예외의 정의와 발생
      </a>
</div>
</div>
<br/>
</div>
<footer class="container-fluid wm-page-content">
<p>Documentation built with <a href="https://www.mkdocs.org/">MkDocs</a> using <a href="https://github.com/gristlabs/mkdocs-windmill">Windmill</a> theme by Grist Labs.</p>
</footer>
<script type="module">import mermaid from "https://unpkg.com/mermaid@10.4.0/dist/mermaid.esm.min.mjs";
mermaid.initialize({});</script></body>
</html>